aws/resource_aws_route53_resolver_rule.go

```diff
@@ -0,0 +1,293 @@
+package aws
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53resolver"
+)
+
+const (
+	route53ResolverRuleStatusDeleted = "DELETED"
+)
+
+func resourceAwsRoute53ResolverRule() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsRoute53ResolverRuleCreate,
+		Read:   resourceAwsRoute53ResolverRuleRead,
+		Update: resourceAwsRoute53ResolverRuleUpdate,
+		Delete: resourceAwsRoute53ResolverRuleDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(10 * time.Minute),
+			Update: schema.DefaultTimeout(10 * time.Minute),
+			Delete: schema.DefaultTimeout(10 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"domain_name": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: suppressRoute53ZoneNameWithTrailingDot,
+				ValidateFunc:     validation.StringLenBetween(1, 256),
```

Comment:
`domain_name = "."` is failing `ValidateFunc` with
```
aws_route53_resolver_rule.dot: error creating Route 53 Resolver rule: InvalidParameter: 1 validation error(s) found.
- minimum field size of 1, CreateResolverRuleInput.DomainName.
```
I also tested with `domain_name = ".."` to see if a `.` was being dropped, but that fails with an entirely reasonable
```
* aws_route53_resolver_rule.dot: error creating Route 53 Resolver rule: InvalidParameterException: [RSLVR-00711] Invalid domain name```

Reply:
Yes, with `domain_name = "."` `terraform plan` shows no `domain_name` attribute.
I guess that's because according to `suppressRoute53ZoneNameWithTrailingDot()` `"."` and `""` are the same.

Reply:
Can you update `suppressRoute53ZoneNameWithTrailingDot()` for this case then, since `"."` is valid here, and in fact explicitly suggested as a catch-all rule called ForwardAll in AWS training materials?

I don't think changing the func will impact the other R53 code that uses the same DiffSuppressFunc, because it's for Diff Suppression, not Validation. Also, AWS will happily throw `Invalid domain name` for cases when `"."` can't be used.

Reply:
If I fix that _DiffSuppressFunc_ i have verified that we can create rule for `domain_name = "."`:

```console
$ terraform apply
aws_route53_resolver_rule.example: Creating...
  arn:          "" => "<computed>"
  domain_name:  "" => "."
  name:         "" => "terraform-testacc-r53-resolver-foofoo"
  owner_id:     "" => "<computed>"
  rule_type:    "" => "SYSTEM"
  share_status: "" => "<computed>"
aws_route53_resolver_rule.example: Still creating... (10s elapsed)
aws_route53_resolver_rule.example: Creation complete after 11s (ID: rslvr-rr-00000000000000000)

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

Reply:
The other user of `suppressRoute53ZoneNameWithTrailingDot()`, `aws_route43_zone` has the same problem with `name = "."` but nobody would have spotted this as you can't create such a Route 53 Zone:

```console
$ terraform apply
aws_route53_zone.test: Creating...
  comment:        "" => "Managed by Terraform"
  force_destroy:  "" => "false"
  name:           "" => "."
  name_servers.#: "" => "<computed>"
  vpc_id:         "" => "<computed>"
  vpc_region:     "" => "<computed>"
  zone_id:        "" => "<computed>"
Error applying plan:

1 error(s) occurred:

* aws_route53_zone.test: 1 error(s) occurred:

* aws_route53_zone.test: error creating Route53 Hosted Zone: InvalidDomainName: .
	status code: 400, request id: d9c6051b-41da-11e9-83dd-b3dabfee8d3a
```

Reply:
`suppressRoute53ZoneNameWithTrailingDot()` fixed and unit tests added.
Acceptance tests re-run:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsRoute53ResolverRule_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsRoute53ResolverRule_ -timeout 120m
=== RUN   TestAccAwsRoute53ResolverRule_basic
=== PAUSE TestAccAwsRoute53ResolverRule_basic
=== RUN   TestAccAwsRoute53ResolverRule_tags
=== PAUSE TestAccAwsRoute53ResolverRule_tags
=== RUN   TestAccAwsRoute53ResolverRule_updateName
=== PAUSE TestAccAwsRoute53ResolverRule_updateName
=== RUN   TestAccAwsRoute53ResolverRule_forward
=== PAUSE TestAccAwsRoute53ResolverRule_forward
=== CONT  TestAccAwsRoute53ResolverRule_basic
=== CONT  TestAccAwsRoute53ResolverRule_forward
=== CONT  TestAccAwsRoute53ResolverRule_updateName
=== CONT  TestAccAwsRoute53ResolverRule_tags
--- PASS: TestAccAwsRoute53ResolverRule_basic (37.16s)
--- PASS: TestAccAwsRoute53ResolverRule_updateName (57.16s)
--- PASS: TestAccAwsRoute53ResolverRule_tags (57.59s)
--- PASS: TestAccAwsRoute53ResolverRule_forward (384.63s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	384.647s
```